infrastructure/smart-contracts/contracts/Achievements.sol

```diff
@@ -0,0 +1,147 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.9;
+
+import "@openzeppelin/contracts/token/ERC1155/ERC1155.sol";
+import "@openzeppelin/contracts/access/IAccessControl.sol";
+
+/**
+ * @title Achievements Contract
+ * @dev An ERC1155 contract that represents game achievements.
+ */
+contract Achievements is ERC1155 {
+
+    //constants
+    bytes32 public constant MINT_ROLE = keccak256("MINT_ROLE");
+
+    //state
+    IAccessControl public factoryAddress; //factory proxy to check access control
+    mapping(uint256 => bool) public definedTokens; //tokenId => exists
+    uint256[] private _tokenIdList; //array of token ids that exist (ordered by mint order)
+    uint256 public tokenIdCount; //counter for unique token ids that exist
+    uint256 public totalSupply; //counter for total tokens minted on this contract
+    mapping(uint256 => uint256) public totalSupplyById; //tokenId => tokenCount
+
+    /**
+     * @dev Achievements constructor.
+     * @param _factoryAddress Address of the factory contract that produced this contract.
+     * @param _uri URI string pointing to metadata (e.g. https://metadata.xai.games/id.json)
+     */
+    constructor(address _factoryAddress, string memory _uri) ERC1155(_uri) {
+        factoryAddress = IAccessControl(_factoryAddress);
+    }
+
+    /**
+     * @dev Restricts caller to MINT_ROLE holder in factory contract.
+     */
+    modifier onlyMintRole() {
+        require(factoryAddress.hasRole(MINT_ROLE, address(msg.sender)), "caller does not have MINT_ROLE");
+        _;
+    }
+
+    /**
+     * @dev Mints a single token to the recipient address. Only callable by MINT_ROLE holder.
+     * @param to Address to receive the newly minted token.
+     * @param id Token ID of token to mint.
+     * @param data Arbitrary data field.
+     */
+    function mint(address to, uint256 id, bytes memory data) public onlyMintRole {
+        require(balanceOf(to, id) == 0, "address has non-zero token balance");
+        
+        if (!definedTokens[id]) {
```

Comment:
It seems we never set the `definedTokens[id]` to true, is this planning in another ticket ?

Reply:
Good catch. Fixed